stock_putaway_product/models/product_putaway.py

```diff
@@ -0,0 +1,76 @@
+# © 2016 Jos De Graeve - Apertoso N.V. <Jos.DeGraeve@apertoso.be>
+# © 2016 Carlos Dauden - Tecnativa <carlos.dauden@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+
+
+class ProductPutaway(models.Model):
+    _inherit = 'product.putaway'
+
+    @api.model
+    def _get_putaway_options(self):
+        ret = super(ProductPutaway, self)._get_putaway_options()
+        return ret + [('per_product', 'Fixed per product location')]
+
+    product_location_ids = fields.One2many(
+        comodel_name='stock.product.putaway.strategy',
+        inverse_name='putaway_id',
+        string='Fixed per product location',
+        copy=True)
+    method = fields.Selection(selection=_get_putaway_options)
+
+    @api.multi
+    def get_product_putaway_strategies(self, product):
+        """ Get linked product putaway strategy that contain fixed location.
+        product.product_putaway_ids is the fasted way to get strategy
+        especially when we have thousand of products.
+        """
+        self.ensure_one()
+        strategy = product.product_putaway_ids.filtered(
+            lambda strategy: (strategy.putaway_id == self))
+        if not strategy:
+            strategy = product.product_tmpl_id.product_putaway_ids.filtered(
```

Comment:
I think that a fallback variant is still required. As if, for whatever reason, Not product, nor it's template has a strategy -> this will fail. And this is pretty much possible as we saw during fixing of the tests.

Reply:
It indeed makes sense. Here is what I suggest with 83d1bfa:
* if ``method == 'per_product'`` and if a strategy has been found:
```return strategy.fixed_location_id.id``` (as formerly)
* else fall back to the parent class ``putaway_apply`` function

See https://github.com/asaunier/stock-logistics-warehouse/blob/83d1bfa131da574cc0c06b3dbc4d3de0ef82ba3c/stock_putaway_product/models/product_putaway.py#L38-L44